src/sagemaker_training/torch_distributed.py

```diff
@@ -0,0 +1,158 @@
+        processes_per_host,
+        master_hostname,
+        hosts,
+        current_host,
+        network_interface_name,
+        current_instance_type="ml.trn1.2xlarge",
+    ):
+        """Initialize a Native PT Launcher, which is responsible for executing
+        the user entry point within a process.
+
+        Args:
+            user_entry_point (str): The name of the user entry point.
+            args ([str]): A list of arguments to include when executing the entry point.
+            env_vars (dict(str,str)): A dictionary of environment variables.
+        """
+        super(TorchDistributedRunner, self).__init__(
+            user_entry_point, args, env_vars, processes_per_host
+        )
+
+        self._master_hostname = master_hostname
+        self._hosts = hosts
+        self._current_host = current_host
+        self._network_interface_name = network_interface_name
+        self._rdzv_backend = RDZV_BACKEND
+        self._instance_type = current_instance_type
+
+    def _setup(self):
+        logger.info("Starting distributed training through torchrun")
+        os.environ["FI_EFA_USE_DEVICE_RDMA"] = "1"
+        os.environ["FI_PROVIDER"] = "efa"
+        os.environ["NCCL_SOCKET_IFNAME"] = str(self._network_interface_name)
+
+    def _python_command(self):  # pylint: disable=no-self-use
+        return python_executable()
+
+    def _create_command(self):
+        """
+        Based on the number of hosts, torchrun command differs.
+        Currently the elasticity feture of torchrun is not yet supported.
+        """
+        entrypoint_type = _entry_point_type.get(environment.code_dir, self._user_entry_point)
+
+        if entrypoint_type is _entry_point_type.PYTHON_PACKAGE:
+            raise errors.ClientError(
+                "Distributed Training through torch_distributed is not supported \
+                    for Python packages"
+                "Please use a python script as the entry-point"
+            )
+
+        instance_type = self._instance_type
+
+        if entrypoint_type is _entry_point_type.PYTHON_PROGRAM:
+            num_hosts = len(self._hosts)
+            torchrun_cmd = []
+
+            node_options = [
+                TORCH_DISTRIBUTED_MODULE,
+                "--nnodes",
+                str(num_hosts),
+                "--nproc_per_node",
+                str(self._processes_per_host),
+                "--master_addr",
+                str(self._master_hostname),
+                "--master_port",
+                MASTER_PORT,
+                "--node_rank",
+                str(self._hosts.index(self._current_host)),
+            ]
+            torchrun_cmd += node_options
+
+            # match = re.match(r"^ml[\._]([a-z\d]+)\.?\w*$", instance_type)
+
+            torchrun_cmd.append(str(self._user_entry_point))
+            torchrun_cmd += self._args
+            return torchrun_cmd
+        else:
+            raise errors.ClientError("Unsupported entry point type for torch_distributed")
```

Comment:
Are we planning to add validation for this in Python SDK?

Reply:
Added these checks in the PySDK PR https://github.com/aws/sagemaker-python-sdk/pull/3424